src/bigint.rs

```diff
@@ -467,146 +522,303 @@ impl Unsigned for BigUint {}
 
 forward_all_binop_to_val_ref_commutative!(impl Add for BigUint, add);
 
-impl<'a> Add<&'a BigUint> for BigUint {
+fn add2(a: &mut [BigDigit], b: &[BigDigit]) {
+    let mut b_iter = b.iter();
+    let mut carry = 0;
+
+    for ai in a.iter_mut() {
+        if let Some(bi) = b_iter.next() {
+            *ai = adc(*ai, *bi, &mut carry);
+        } else if carry != 0 {
+            *ai = adc(*ai, 0, &mut carry);
+        } else {
+            break;
+        }
+    }
+
+    assert!(b_iter.next() == None && carry == 0);
+}
+
+impl<'a> Add<&'a [BigDigit]> for BigUint {
```

Comment:
On Mon, Nov 30, 2015 at 03:39:34PM -0800, Josh Stone wrote:
> > +    let mut carry = 0;
> > +
> > +    for ai in a.iter_mut() {
> > +        if let Some(bi) = b_iter.next() {
> > +            *ai = adc(*ai, *bi, &mut carry);
> > +        } else if carry != 0 {
> > +            *ai = adc(*ai, 0, &mut carry);
> > +        } else {
> > +            break;
> > +        }
> > +    }
> > +
> > +    assert!(b_iter.next() == None && carry == 0);
> > +}
> > +
> > +impl<'a> Add<&'a [BigDigit]> for BigUint {
> 
> This new impl is visible in the public API, which is undesirable.  AFAICS this only used by the normal Add implementation, so let's keep it as one.

Argh, I didn't think about that. Is it possible to keep it private? I know it's
not used in this patch series, but I added it because I'm really trying to get
away from using add2/sub2 directly - due to a rather embarassing bug I found the
other night (in sub_sign), which never would've slipped by except for all the
boilerplate from using sub2 directly.

kind of an aside - btw, do you work on the compiler at all? the other usage of
add2() that I haven't been able to get away from is all being used to optimize
code like

acc += b << shift

If we a shl_accumulate method, and some way to teach the compiler/optimizer to
replace (maybe think of it as a type of specialization?) expressons like that
with our shl_accumulate method - we could write a lot of code more naturally,
and the users of this library would benefit too.

You could think of it as generalizing peephole optimizations, I think. Once
you're in SSA form, everything is of the form

f(g(x, y), z)

hand waving a bit and writing it differently than usual - but every value is a
pointer to the thing that produces it, right? So just pretending every value is
a function call.

So I think this might end up being fairly simple: the idea is, if I implement
functions f and g, I can provide a function f_g (or whatever) and tell the
compiler that expressions of the form f(g(x, y), z) can be replaced with calls
to f_g(x, y, z) - which in this case is this shl_accumulate method.

Make any sense? I mean - the compiler could do all these optimizations if the
code was working on primitive types, but because we're doing math with our
special fancy types none of those optimizations are available. It'd be really
cool to be able to write all this math code in a sane natural way.